### ISSUE TITLE
reproduce the results of DARPA 

### ISSUE THREAD

**Issue by DarrenWong** [leggedrobotics/DeLORA#19]
Dear officer,

Thanks for sharing a great job.I am running the code with the experiments DARPA SubT Challenge Urban Circuit in the paper _Self-supervised Learning of LiDAR Odometry for Robotic Applications_

First I trained a network base on the Alpha course for around 80 epochs with a loss of 0.077586
![image](https://user-images.githubusercontent.com/5771106/164954114-c40b3cc6-a813-49bd-9ecc-ae5520026162.png)

Then I use the Beta to infer the results
![darpa_beta_map_darpa_00_2d](https://user-images.githubusercontent.com/5771106/164954311-97d2c792-23f6-4257-9818-29e39f9aa674.png)
![darpa_beta_map_darpa_00_3d](https://user-images.githubusercontent.com/5771106/164954316-1d442bb4-efeb-41b3-add5-9266a4251419.png)

seems not very close to the paper 

![image](https://user-images.githubusercontent.com/5771106/164954170-6848215b-9813-4769-b9db-40eec99146a4.png)

I also try to use a LOAM-like method, it can output a map so I think the data I am using is correct.
![result_legoloam](https://user-images.githubusercontent.com/5771106/164954184-a7350911-0409-446f-8006-02cb0a6c3275.png)

My question is,
1. Does the trained model is not good? Can you share your model on DARPA for testing?
2. Do I need to add the mapping module of LOAM before evaluating the results of DARPA?

**Comment by nubertj** [comment]
HI @DarrenWong,
Sorry for the late reply!
I am super busy these days with multiple deliverables and ICRA 2022 coming up.

1) I will look into it after ICRA, promise! I will retrain a model for you and send you the instructions on how to reproduce the results from the paper.
2) Yes, the results in the paper were achieved with the LOAM mapping module. Only the scan2scan has been replaced.

Best,
Julian

**Comment by DarrenWong** [comment]
@nubertj Thanks for your kind response. Look forward to your updates and the latest work in ICRA 2022. You can send me by email: darren-f.huang@connect.polyu.hk

**Comment by GavyndotLee** [comment]
Hi @nubertj . I encountered a similar problem. Whether or not I modify the parameters in the config, the trained network always makes mistakes in predicting rotation. I want to locate the problem, and can you provide model parameters with the experiments DARPA SubT Challenge Urban Circuit?
![fig](https://user-images.githubusercontent.com/115397140/201611592-8f192ed2-bae9-414e-81ed-97b13e0195f2.png)


Thanks for your help!

**Comment by nubertj** [comment]
HI @GavyndotLee,
**How did you generate the map from the screenshot? 
1. If it is the validation set, I assume you already integrated into the scan2map module?
2. If not, I assume it is the training set?**

Independent Note: 
----------------------------------------------------
I can share the model, but I did some pretty big adjustments in the code before open sourcing it, so I would need to retrain the model. As I am currently doing an internship in the USA it is not so easy to find the time for that, but I can try.

But in general to make sure: Clearly the generalization capability of the network depends on the amount of training data you use (as always). Also the training procedure is not the most stable one, so the performance (in particular on kitti) can vary quite a bit from epoch to epoch. 
If I really only trained on the Darpa SubT run, e.g. the alpha course, the performance on the alpha course was pretty good, but the performance on the beta course was not great, when only looking at the scan2scan output. For this reason, but also for the reason that scan2scan alone barely works in practice (also the LOAM scan2scan produces a terrible overall trajectory), we integrated it with the san2map.

If nice generalization is needed, I fear that simply more data needs to be used for training, compared to the output on kitti.

Best,
Julian

**Comment by GavyndotLee** [comment]
HI @nubertj,
I'm sorry to bother you, but I do need your help.
I have integrated scan2scan into the scan2map module of LOAM based on frame convention in ROS (X-forward, Y-left, and Z-up). The map as shown was generated from the validation set.

In addition, even if I use the training set, the prediction is not satisfactory, as shown in the following figure. 
![Fig1](https://user-images.githubusercontent.com/115397140/202100158-e47ca519-5b3d-4426-b608-938524564dbf.png)

To eliminate the problem of scan2map, I supplement the running results of the LOAM method on the training set and validation set.
![Fig2](https://user-images.githubusercontent.com/115397140/202100322-ef0fe035-4dd2-4eb3-a38c-c3bcaf435178.png)
![Fig3](https://user-images.githubusercontent.com/115397140/202100328-a8d128d2-c05e-4c54-9539-1d2909a7b3ca.png)

**Comment by gqding** [comment]
Hi,

I was able to run the script "run_rosnode.py", and I also run LOAM and "rosbag play", it seems the rosenode is predicting the odometry results of delora
![image](https://user-images.githubusercontent.com/35989403/224221017-720d6a03-efb6-4a90-b1e4-f1ba3d574e64.png)

However, when I run LOAM, it only show the current scans in the rviz. 
![image](https://user-images.githubusercontent.com/35989403/224221691-bc27785d-99aa-421d-820a-ee212261cb93.png)

Could you please let me know how to obtain the whole map with delora odometry and how to save the odom+mapping pose results?

Bests.
